### PR TITLE
Some developer documentation for memory bounds checking

### PR DESCRIPTION
**Pull request by danpoe**
- [ ] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- n/a Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Comment by majakusber** [comment]
# [Codecov](https://codecov.io/gh/diffblue/cbmc/pull/5228?src=pr&el=h1) Report
> Merging [#5228](https://codecov.io/gh/diffblue/cbmc/pull/5228?src=pr&el=desc) into [develop](https://codecov.io/gh/diffblue/cbmc/commit/cccdfd9ba3c1033b3634823884f21dbea030bf33&el=desc) will **decrease** coverage by `1.55%`.
> The diff coverage is `n/a`.

[Impacted file tree graph](https://codecov.io/gh/diffblue/cbmc/pull/5228?src=pr&el=tree)

```diff
@@             Coverage Diff             @@
##           develop    #5228      +/-   ##
===========================================
- Coverage    67.43%   65.88%   -1.56%     
===========================================
  Files         1157     1157              
  Lines        95354    95354              
===========================================
- Hits         64306    62828    -1478     
- Misses       31048    32526    +1478     
```

| Flag | Coverage Δ | |
|---|---|---|
| #cproversmt2 | `?` | |
| #regression | `63.96% <ø> (ø)` | :arrow_up: |
| #unit | `31.95% <ø> (ø)` | :arrow_up: |

| [Impacted Files](https://codecov.io/gh/diffblue/cbmc/pull/5228?src=pr&el=tree) | Coverage Δ | |
|---|---|---|
| [src/solvers/smt2/letify.cpp](https://codecov.io/gh/diffblue/cbmc/pull/5228/diff?src=pr&el=tree#diff-c3JjL3NvbHZlcnMvc210Mi9sZXRpZnkuY3Bw) | `0.00% <0.00%> (-100.00%)` | :arrow_down: |
| [src/solvers/smt2/smt2\_dec.h](https://codecov.io/gh/diffblue/cbmc/pull/5228/diff?src=pr&el=tree#diff-c3JjL3NvbHZlcnMvc210Mi9zbXQyX2RlYy5o) | `0.00% <0.00%> (-100.00%)` | :arrow_down: |
| [src/solvers/floatbv/float\_bv.h](https://codecov.io/gh/diffblue/cbmc/pull/5228/diff?src=pr&el=tree#diff-c3JjL3NvbHZlcnMvZmxvYXRidi9mbG9hdF9idi5o) | `0.00% <0.00%> (-100.00%)` | :arrow_down: |
| [src/solvers/smt2/smt2irep.cpp](https://codecov.io/gh/diffblue/cbmc/pull/5228/diff?src=pr&el=tree#diff-c3JjL3NvbHZlcnMvc210Mi9zbXQyaXJlcC5jcHA=) | `0.00% <0.00%> (-88.00%)` | :arrow_down: |
| [src/solvers/floatbv/float\_bv.cpp](https://codecov.io/gh/diffblue/cbmc/pull/5228/diff?src=pr&el=tree#diff-c3JjL3NvbHZlcnMvZmxvYXRidi9mbG9hdF9idi5jcHA=) | `0.00% <0.00%> (-74.37%)` | :arrow_down: |
| [src/solvers/smt2/smt2\_dec.cpp](https://codecov.io/gh/diffblue/cbmc/pull/5228/diff?src=pr&el=tree#diff-c3JjL3NvbHZlcnMvc210Mi9zbXQyX2RlYy5jcHA=) | `0.00% <0.00%> (-67.70%)` | :arrow_down: |
| [src/solvers/smt2/letify.h](https://codecov.io/gh/diffblue/cbmc/pull/5228/diff?src=pr&el=tree#diff-c3JjL3NvbHZlcnMvc210Mi9sZXRpZnkuaA==) | `33.33% <0.00%> (-66.67%)` | :arrow_down: |
| [src/solvers/smt2/smt2\_conv.h](https://codecov.io/gh/diffblue/cbmc/pull/5228/diff?src=pr&el=tree#diff-c3JjL3NvbHZlcnMvc210Mi9zbXQyX2NvbnYuaA==) | `36.36% <0.00%> (-63.64%)` | :arrow_down: |
| [src/solvers/smt2/smt2\_conv.cpp](https://codecov.io/gh/diffblue/cbmc/pull/5228/diff?src=pr&el=tree#diff-c3JjL3NvbHZlcnMvc210Mi9zbXQyX2NvbnYuY3Bw) | `19.81% <0.00%> (-37.43%)` | :arrow_down: |
| [src/solvers/flattening/pointer\_logic.cpp](https://codecov.io/gh/diffblue/cbmc/pull/5228/diff?src=pr&el=tree#diff-c3JjL3NvbHZlcnMvZmxhdHRlbmluZy9wb2ludGVyX2xvZ2ljLmNwcA==) | `71.21% <0.00%> (-18.19%)` | :arrow_down: |
| ... and [17 more](https://codecov.io/gh/diffblue/cbmc/pull/5228/diff?src=pr&el=tree-more) | |

------

[Continue to review full report at Codecov](https://codecov.io/gh/diffblue/cbmc/pull/5228?src=pr&el=continue).
> **Legend** - [Click here to learn more](https://docs.codecov.io/docs/codecov-delta)
> `Δ = absolute <relative> (impact)`, `ø = not affected`, `? = missing data`
> Powered by [Codecov](https://codecov.io/gh/diffblue/cbmc/pull/5228?src=pr&el=footer). Last update [cccdfd9...31e9f50](https://codecov.io/gh/diffblue/cbmc/pull/5228?src=pr&el=lastupdated). Read the [comment docs](https://docs.codecov.io/docs/pull-request-comments).

File: doc/architectural/memory-bounds-checking.md [hunk 1]
```diff
@@ -0,0 +1,85 @@
+\ingroup module_hidden
+\page memory-bounds-checking Memory Bounds Checking
+
+cbmc provides means to verify that pointers are within the bounds of (statically
+or dynamically) allocated memory blocks. When the option `--pointer-check` is
+used, cbmc checks the safety of each pointer dereference in the program.
+Similary, the primitive `__CPROVER_r_ok(pointer, size)` returns true when it is
```
Comment (xbauch) [inline doc/architectural/memory-bounds-checking.md]:
```suggestion
Similarly, the primitive `__CPROVER_r_ok(pointer, size)` returns true when it is
```

File: doc/architectural/memory-bounds-checking.md [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+  __CPROVER_bool record_malloc = __VERIFIER_nondet___CPROVER_bool();
+  __CPROVER_malloc_object = record_malloc ? malloc_res : __CPROVER_malloc_object;
+  __CPROVER_malloc_size = record_malloc ? malloc_size : __CPROVER_malloc_size;
+
+  return malloc_res;
+}
+```
+
+Both internal variables `__CPROVER_malloc_object` and `__CPROVER_malloc_size`
+are initialised to 0 in the `__CPROVER_initialize()` function of a goto program.
```
Comment (xbauch) [inline doc/architectural/memory-bounds-checking.md]: I think at some point it was decided that we stick to the Oxford spelling in documentation, i.e. "are initialized ..".

**Comment by majakusber** [comment]
Oh no, my face on the codecov again :woman_shrugging: If it happens third time, I'll look into it. I promise I'm not monitoring your work @xbauch :sweat_smile:

**Comment by xbauch** [comment]
> Oh no, my face on the codecov again  If it happens third time, I'll look into it. I promise I'm not monitoring your work @xbauch

Not my work in this case; it seems you're somehow responsible for the coverage of cprover documentation.

**Comment by majakusber** [comment]
The second codecov report is correctly by `codecov-io`, and that one actually seems correct - documentation should show no change to coverage I guess. So I'm hoping someone/something fixed it (just as mysteriously as it started using my face) :crossed_fingers:

**Review by peterschrammel — APPROVED** [review]

**Review by xbauch — APPROVED** [review]
Looks good.

**Review by martin-cs — APPROVED** [review]
Don't suppose you want to document how memory leaks and use-after-free are handled do you? :-)

**Pull request merged**